### PR TITLE
Logistic#1

### PR DESCRIPTION
**Pull request by MuhamedEssam**
have a look on this please and comments mean questions

### CODE REVIEW

File: client/tinychain/ml/mlr.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    
+    def cost_func( self, labels ,preds):
+        '''
+        I try to apply Mean Squared Error
+        '''
+        cost = Tensor.sum((labels - preds)**2)/labels.len()
+        return cost
+
+    def gradient_descent(self , inputs , weights, labels):        
+        d_weights = 2/inputs.shape[0]*(einsum("ij,ki->kj", weights, einsum("ij,ki->kj", inputs, inputs.transpose()))-
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:43]: Again, there's already a `GradientDescent` optimizer that you can use for this

File: client/tinychain/ml/mlr.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        return new_weights
+
+    def fit(self, X_train, y_train,iteration):
+        X_train_ = Tensor.concatenate(1,X_train,axis=1) 
+        y_train_ = y_train
+        n_samples, n_features = X_train_.shape[1],X_train_.shape[0]
+        '''
+        Can i create zeros tc tensor  or i have to create it in numpy then convert?
+        '''       
+        weights = np.zeros((n_features + 1, 1))
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:59]: It's OK to use NumPy in tests but the client should never ever depend on NumPy. You can use `Dense.zeros` here instead if you need.

Comment (haydnv) [reply client/tinychain/ml/mlr.py:59]: Although I think you should consider using a `DNNLayer` with the `Identity` activation function in order to avoid duplicating logic

File: client/tinychain/ml/mlr.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        n_samples, n_features = X_train_.shape[1],X_train_.shape[0]
+        '''
+        Can i create zeros tc tensor  or i have to create it in numpy then convert?
+        '''       
+        weights = np.zeros((n_features + 1, 1)) 
+        weights_= Tensor.load(weights.shape,tc.F32,weights.flatten().tolist())        
+        
+ 
+        @tc.get_op
+        def loop(iteration: tc.Number) -> tc.Int:
```
Comment (haydnv) [inline client/tinychain/ml/mlr.py:64]: this logic also seems very similar to `GradientDescent.train`

File: tests/client/test_machine.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+Dense = tc.tensor.Dense
+
+ENDPOINT = "/transact/hypothetical"
+LEARNING_RATE = tc.F32(0.01)
+MAX_ITERATIONS = 10
+NUM_EXAMPLES = 20
+
+
+class Logistic_Regression():
```
Comment (haydnv) [inline tests/client/test_machine.py:17]: I think it would be simpler to make this a function called maybe `regress` or `multi_regress`. That way you don't have to worry about initialization, inheritance, etc.

File: tests/client/test_machine.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import numpy as np
+import tinychain as tc
+import unittest
+from sklearn.datasets import load_breast_cancer
```
Comment (haydnv) [inline tests/client/test_machine.py:4]: I think it would be easier to debug using a small dataset, e.g. inputs `[0, 1, 2, 3]` and labels `[1, 3, 5, 7]` to check that regression correctly fits `y = 2x + 1`

File: tests/client/test_ml.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+
```
Comment (haydnv) [inline tests/client/test_ml.py:1]: I think maybe you didn't mean to include this file?

**Review by haydnv — CHANGES_REQUESTED** [review]

File: client/tinychain/ml/mlr.py (hunk 2 above)
Comment (MuhamedEssam) [reply client/tinychain/ml/mlr.py:59]: thats wrong file i uploaded it before in another pull request dunno why its here

File: tests/client/test_machine.py (hunk 4 above)
Comment (MuhamedEssam) [reply tests/client/test_machine.py:17]: can you explain more ?  does name affect anything ?

**Comment by haydnv** [comment]
Superceded by #115
